safaribooks.py

```diff
@@ -899,17 +899,20 @@ def _thread_download_images(self, url):
         self.display.state(len(self.images), self.images_done_queue.qsize())
 
     def _start_multiprocessing(self, operation, full_queue):
-        if len(full_queue) > 5:
-            for i in range(0, len(full_queue), 5):
-                self._start_multiprocessing(operation, full_queue[i:i + 5])
-
-        else:
-            process_queue = [Process(target=operation, args=(arg,)) for arg in full_queue]
-            for proc in process_queue:
-                proc.start()
-
-            for proc in process_queue:
-                proc.join()
+            try:
+                if len(full_queue) > 5:
+                    for i in range(0, len(full_queue), 5):
+                        self._start_multiprocessing(operation, full_queue[i:i + 5])
```

Comment:
**Suggestion:** The recursive call to _start_multiprocessing within the same function can lead to a stack overflow if the queue size is large. Consider using a while loop instead.
```suggestion
            def _start_multiprocessing(self, operation, full_queue):
                try:
                    while len(full_queue) > 5:
                        chunk = full_queue[:5]
                        full_queue = full_queue[5:]
                        process_queue = [Process(target=operation, args=(arg,)) for arg in chunk]
                        for proc in process_queue:
                            proc.start()
                        for proc in process_queue:
                            proc.join()
```